File: Java/src/main/java/Solution501_600/Solution506.java
package Solution501_600;

import java.util.Arrays;
import java.util.HashMap;
import java.util.stream.Collectors;

/**
 * @author windows
 */
public class Solution506 {
    public String[] findRelativeRanks(int[] score) {
        int[] arr = new int[score.length];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = score[i];
        }
        Arrays.sort(arr);
        int len = score.length;
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < len; i++) {
            map.put(arr[i], len - i);
        }
        String[] strings = new String[score.length];
        for (int i = 0; i < len; i++) {
            if (map.get(score[i]) == 1) {
                strings[i] = "Gold Medal";
            }else if (map.get(score[i]) == 2){
                strings[i] = "Silver Medal";
            }else if(map.get(score[i]) == 3){
                strings[i] = "Bronze Medal";
            }else{
                strings[i] = String.valueOf(map.get(score[i]));
            }
        }
        return strings;
    }


    public static void main(String[] args) {
        Solution506 solution506 = new Solution506();
        System.out.println(Arrays.toString(solution506.findRelativeRanks(new int[]{5, 4, 3, 2, 1})));
    }
}

File: C++/src/Solution501_600/Solution506.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution506.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <unordered_map>
#include <algorithm>

//输入：score = [10,3,8,9,4]
//输出：["Gold Medal","5","Bronze Medal","Silver Medal","4"]
//解释：名次为 [1st, 5th, 3rd, 2nd, 4th]
vector<string> Solution506::findRelativeRanks(vector<int> &score) {
    int n = score.size();
    unordered_map<int, string> rank_map;
    vector<int> sort_score = score;
    sort(sort_score.rbegin(), sort_score.rend());
    rank_map[sort_score[0]] = "Gold Medal";
    if (n > 1) rank_map[sort_score[1]] = "Silver Medal";
    if (n > 2) rank_map[sort_score[2]] = "Bronze Medal";
    for (int i = 3; i < n; i++) {
        rank_map[sort_score[i]] = to_string(i + 1);
    }
    vector<string> res;
    for (int i = 0; i < n; ++i) {
        res.push_back(rank_map[score[i]]);
    }
    return res;
}
